src/catalog/src/durable/upgrade/v58_to_v59.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::durable::upgrade::MigrationAction;
+use crate::durable::upgrade::{objects_v58 as v58, objects_v59 as v59};
+use mz_proto::wire_compatible;
+use mz_proto::wire_compatible::WireCompatible;
+
+wire_compatible!(v58::ClusterReplicaKey with v59::ClusterReplicaKey);
+wire_compatible!(v58::ClusterId  with v59::ClusterId);
+wire_compatible!(v58::RoleId  with v59::RoleId);
+wire_compatible!(v58::ReplicaLogging with v59::ReplicaLogging);
+
+/// In v59, we add a pending to cluster replica ManagedLocation
+pub fn upgrade(
+    snapshot: Vec<v58::StateUpdateKind>,
+) -> Vec<MigrationAction<v58::StateUpdateKind, v59::StateUpdateKind>> {
+    snapshot
+        .into_iter()
+        .filter_map(|update| {
+            let v58::state_update_kind::Kind::ClusterReplica(
+                v58::state_update_kind::ClusterReplica { key, value },
+            ) = update.kind.as_ref().expect("missing field")
+            else {
+                return None;
+            };
+            let (logging, managed_config) = match value {
+                Some(v) => match v.config.as_ref() {
+                    Some(c) => {
+                        let logging = c.logging.clone();
+                        match c.location.as_ref() {
+                            Some(l) => match l {
+                                v58::replica_config::Location::Managed(ml) => (logging, ml.clone()),
+                                _ => return None,
+                            },
+                            _ => return None,
+                        }
+                    }
+                    _ => return None,
+                },
+                _ => return None,
+            };
```

Comment:
```suggestion
            let c = value?.config.as_ref()?;
            let v58::replica_config::Location::Managed(ml) = c.location.as_ref()? else {
                return None;
            };
            let logging = c.logging.clone();
            let managed_config = ml.clone();
```

Untested but I believe works. `?` supposedly knows about `None`.

Reply:
Awesome!!!
worked with a small modification.
`let c = value.as_ref()?.config.as_ref()?;`